File: KonovalovIlya/chat/src/main/java/ru/turbogoose/core/MessageLog.java
package ru.turbogoose.core;

import java.util.Collections;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

public class MessageLog implements Iterable<Message> {
    // sorted by timestamp and then by id
    private final List<Message> messages = Collections.synchronizedList(new LinkedList<>());

    public void addMessage(Message message) {
        int indexToInsert;
        for (indexToInsert = 0; indexToInsert < messages.size(); indexToInsert++) {
            Message msgInLog = messages.get(indexToInsert);
            if (!(message.getTimestamp().after(msgInLog.getTimestamp()) ||
                    message.getTimestamp().equals(msgInLog.getTimestamp()) &&
                    message.getId().compareTo(msgInLog.getId()) < 0)) {
                break;
            }
        }
        messages.add(indexToInsert, message);
    }

    public boolean contains(Message message) {
        for (Message msg : messages) {
            if (msg.equals(message)) {
                return true;
            }
        }
        return false;
    }

    @Override
    public Iterator<Message> iterator() {
        return messages.iterator();
    }
}

File: Peer2PeerChat/MessageLog.cpp
#include "MessageLog.h"
#include <algorithm>

void MessageLog::addMessage(const Message &message) {
  std::lock_guard<std::mutex> lock(mutex_);
  auto it = std::lower_bound(messages_.begin(), messages_.end(), message,
                             [](const Message &a, const Message &b) {
                               return a.getTimestamp() < b.getTimestamp() ||
                                      (a.getTimestamp() == b.getTimestamp() &&
                                       a.getContent() < b.getContent());
                             });
  messages_.insert(it, message);
}

bool MessageLog::contains(const Message &message) const {
  std::lock_guard<std::mutex> lock(mutex_);
  return std::find(messages_.begin(), messages_.end(), message) !=
         messages_.end();
}

std::vector<Message>::const_iterator MessageLog::begin() const {
  return messages_.begin();
}

std::vector<Message>::const_iterator MessageLog::end() const {
  return messages_.end();
}
